tidy3d/plugins/resonance/resonance.py

```diff
@@ -0,0 +1,272 @@
+
+        zinvl = np.exp(1j * np.outer(np.log(zvals) * 1j, np.arange(M + 1)))
+        factor = zinvl @ signal[: M + 1]
+
+        for k in range(J):
+            amplitudes[k] = np.square(eigvecs[:, k] @ factor)
+
+        return amplitudes
+
+    def _find_errors(
+        self,
+        eigvals: np.ndarray[complex],
+        u_matrices: np.ndarray[complex],
+        eigvecs: np.ndarray[complex],
+    ) -> np.ndarray[complex]:
+        """Find the eigenvalue error"""
+        J = len(eigvals)
+        errors = np.zeros(J)
+        for k in range(J):
+            errors[k] = np.linalg.norm(
+                (u_matrices[2] - (eigvals[k] ** 2) * u_matrices[0]) @ eigvecs[:, k]
+            )
+        return errors
+
+    def _iterate(
+        self, signal: np.ndarray[complex], prev_eigvals: np.ndarray[complex]
+    ) -> ResonanceData:
+        """Run a single iteration of the solver."""
+        u_matrices = self._evaluate_matrices(signal, prev_eigvals)
+
+        eigvals, eigvecs = self._solve_gen_eig_prob(u_matrices[1], u_matrices[0], self.rcond)
+
+        errors = self._find_errors(eigvals, u_matrices, eigvecs)
+
+        amplitudes = self._find_amplitudes(signal, prev_eigvals, eigvals, eigvecs)
+
+        return ResonanceData(eigvals=eigvals, errors=errors, complex_amplitudes=amplitudes)
+
+    def run(self, signal: np.ndarray[complex], time_step: float) -> xr.Dataset:
+        """Finds resonances in a time series.
+        Note that the signal should start after the sources have turned off.
+
+        Parameters
+        ----------
+        signal : np.ndarray[complex]
+            Time series to search for resonances
+        time_step : float
+            Time step / sampling rate of the data (in seconds).
+
+        Returns
+        -------
+        :class:`.xr.Dataset`
+            Dataset containing the frequency, decay rate, Q, amplitude, phase,
+            and estimation error of the resonances. Modes with low Q, small
+            amplitude, or high estimation error are likely to be spurious.
+        """
+        J = self.freq_grid
+        if self.freq_grid is None:
+            J = int(time_step * len(signal) * (self.freq_window[1] - self.freq_window[0]) / 2)
+        if J == 0:
+            raise SetupError("Attempting to start with no grid frequencies.")
+        omegas = np.linspace(
+            time_step * self.freq_window[0] * 2 * np.pi,
+            time_step * self.freq_window[1] * 2 * np.pi,
+            J,
+        )
+        eigvals = np.exp(-1j * omegas)
+        prev_num_eigvals = -1
+        for _ in range(1 + 2 * J):
+            resdata = self._iterate(signal, eigvals)
+            if len(resdata.eigvals) == prev_num_eigvals:
+                break
+            prev_num_eigvals = len(resdata.eigvals)
+
+        resonances = self._get_resonance_info(data=resdata, time_step=time_step)
+
+        return resonances.sortby("freq")
+
+    def run_field_time(self, signal: ScalarFieldTimeDataArray) -> xr.Dataset:
```

Comment:
We discussed providing an option for the user to supply several `ScalarFieldTimeDataArray` objects and add their signals together? Perhaps we could consider 
```py
def _process_scalar_field_time(self, signal: ScalarFieldTimeDataArray) -> Tuple[np.ndarray, float]:
```
to strip out the signal data and dt from a scalar field time data array and then
```py
def run_scalar_field_time(self, signals: List[ScalarFieldTimeDataArray) -> XR.Dataset
```
to call this helper function for each `signal` in `signals`, add their results, check the `dt` are all the same, and `run()` on the combined data and dt?

Reply:
I will save this for later too, but I will go ahead and rename `run_field_time` to `run_scalar_field_time`.